pkg/broker/handler/pool/fanout/pool.go

```diff
@@ -0,0 +1,139 @@
+	"github.com/google/knative-gcp/pkg/broker/handler/processors/deliver"
+	"github.com/google/knative-gcp/pkg/broker/handler/processors/fanout"
+	"github.com/google/knative-gcp/pkg/broker/handler/processors/filter"
+)
+
+// SyncPool is the sync pool for fanout handlers.
+// For each broker in the config, it will attempt to create a handler.
+// It will also stop/delete the handler if the corresponding broker is deleted
+// in the config.
+type SyncPool struct {
+	options *pool.Options
+	targets config.ReadonlyTargets
+	pool    sync.Map
+}
+
+// StartSyncPool starts the sync pool.
+func StartSyncPool(ctx context.Context, targets config.ReadonlyTargets, opts ...pool.Option) (*SyncPool, error) {
+	p := &SyncPool{
+		targets: targets,
+		options: pool.NewOptions(opts...),
+	}
+	if err := p.sync(ctx); err != nil {
+		return nil, err
+	}
+	if p.options.SyncSignal != nil {
+		go p.watch(ctx)
+	}
+	return p, nil
+}
+
+func (p *SyncPool) watch(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-p.options.SyncSignal:
+			if err := p.sync(ctx); err != nil {
+				logging.FromContext(ctx).Error("failed to sync handlers pool on watch signal", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (p *SyncPool) sync(ctx context.Context) error {
+	var errs int
+
+	p.pool.Range(func(key, value interface{}) bool {
+		if _, ok := p.targets.GetBrokerByKey(key.(string)); !ok {
+			value.(*handler.Handler).Stop()
+			p.pool.Delete(key)
+		}
+		return true
+	})
+
+	p.targets.RangeBrokers(func(b *config.Broker) bool {
+		bk := config.BrokerKey(b.Namespace, b.Name)
+
+		// There is already a handler for the broker, skip.
+		if _, ok := p.pool.Load(bk); ok {
+			return true
+		}
+
+		projectIDOpt := pubsub.WithProjectIDFromDefaultEnv()
+		if p.options.ProjectID != "" {
+			projectIDOpt = pubsub.WithProjectID(p.options.ProjectID)
+		}
+		ps, err := pubsub.New(ctx,
+			projectIDOpt,
+			pubsub.WithTopicID(b.DecoupleQueue.Topic),
+			pubsub.WithSubscriptionID(b.DecoupleQueue.Subscription),
+		)
+		if err != nil {
+			logging.FromContext(ctx).Error("failed to create pubsub protocol", zap.String("broker", bk), zap.Error(err))
+			errs++
+			return true
+		}
+
+		h := &handler.Handler{
+			Timeout:      p.options.TimeoutPerEvent,
+			PubsubEvents: ps,
+			Processor: processors.ChainProcessors(
+				&fanout.Processor{MaxConcurrency: p.options.MaxConcurrencyPerEvent, Targets: p.targets},
+				&filter.Processor{},
+				&deliver.Processor{Requester: p.options.EventRequester},
+			),
+		}
+		// Start the handler with broker key in context.
+		h.Start(handlerctx.WithBrokerKey(ctx, bk), func(err error) {
+			if err != nil {
+				logging.FromContext(ctx).Error("handler for broker has stopped with error", zap.String("broker", bk), zap.Error(err))
+			} else {
+				logging.FromContext(ctx).Info("handler for broker has stopped", zap.String("broker", bk))
+			}
+			// Make sure the handler is deleted from the pool.
+			p.pool.Delete(h)
```

Comment:
I think `h.Stop()` should be called too.

Reply:
To reach the code here, it means the handler has already stopped.

Reply:
I am talking more about the `err !=  nil` case. If for some reason the code fails to `OpenInbound` for pubsub, we should stop the handler I believe.

Reply:
Reopening this as I still think there is a chance the handler is not stopped.

Reply:
If `OpenInbound` returns with error instantly, then it *should* mean that the pull was not started. Or at least the handler itself should only call `done` when it's already stopped. For safety, I will call `stop` in the handler once OpenInbound returns. This shouldn't have any side effect.